utils.py

```diff
@@ -35,14 +35,15 @@ def cache_response(response, uri=''):
     if 'ETag' not in response.headers:
         remove_response_from_cache(uri)
         return
-    cached_responses[uri] = response
+    database_helper.cache_response(uri, response)
 
 def caching_request_headers(uri):
-    return {'If-None-Match': cached_responses[uri].headers['ETag']} if uri in cached_responses else {}
+    result = database_helper.fetch_cached_response(uri)
+    return {'If-None-Match': result.headers['ETag']} if result is not None else {}
 
 def cached_response(uri):
```

Comment:
if you used fetch_cached_response in database_helper then let's use it here for this function